fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MvPartitionCompensator.java

```diff
@@ -0,0 +1,696 @@
+        for (LogicalScanOperator scanOperator : scanOperators) {
+            if (!SUPPORTED_PARTITION_COMPENSATE_SCAN_TYPES.contains(scanOperator.getOpType())) {
+                continue;
+            }
+            List<ScalarOperator> partitionPredicate = scanOperatorScalarOperatorMap
+                    .computeIfAbsent(Pair.create(scanOperator, isCompensatePartition), x -> {
+                        return isCompensatePartition ? getCompensatePartitionPredicates(columnRefFactory, scanOperator) :
+                                getScanOpPrunedPartitionPredicates(scanOperator);
+                    });
+            if (partitionPredicate == null) {
+                return null;
+            }
+            partitionPredicates.addAll(partitionPredicate);
+        }
+        return partitionPredicates.isEmpty() ? ConstantOperator.createBoolean(true) :
+                Utils.compoundAnd(partitionPredicates);
+    }
+
+    private static List<ScalarOperator> getCompensatePartitionPredicates(ColumnRefFactory columnRefFactory,
+                                                                         LogicalScanOperator scanOperator) {
+        List<ScalarOperator> partitionPredicate = null;
+        if (scanOperator instanceof LogicalOlapScanOperator) {
+            partitionPredicate = compensatePartitionPredicateForOlapScan((LogicalOlapScanOperator) scanOperator,
+                    columnRefFactory);
+        } else if (SUPPORTED_PARTITION_COMPENSATE_EXTERNAL_SCAN_TYPES.contains(scanOperator.getOpType())) {
+            partitionPredicate = compensatePartitionPredicateForExternalTables(scanOperator);
+        } else {
+            return null;
+        }
+        return partitionPredicate;
+    }
+
+
+    private static boolean isSupportedPartitionPredicateForExternalTables(List<PartitionKey> partitionKeys) {
+        for (PartitionKey partitionKey : partitionKeys) {
+            // only support one partition column now.
+            if (partitionKey.getKeys().size() != 1) {
+                return false;
+            }
+            LiteralExpr e = partitionKey.getKeys().get(0);
+            // Only support date/int type
+            if (!(e instanceof DateLiteral || e instanceof IntLiteral)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private static List<ScalarOperator> compensatePartitionPredicateForExternalTables(LogicalScanOperator scanOperator) {
+        try {
+            ScanOperatorPredicates scanOperatorPredicates = scanOperator.getScanOperatorPredicates();
+
+            List<ScalarOperator> partitionPredicates = Lists.newArrayList();
+            Table baseTable = scanOperator.getTable();
+            if (baseTable.isUnPartitioned()) {
+                return partitionPredicates;
+            }
+
+            OperatorType operatorType = scanOperator.getOpType();
+            // only used for hive
+            if (operatorType == OperatorType.LOGICAL_HIVE_SCAN) {
```

Comment:
no other scan types ?

Reply:
as comments: it's not the same meaning when selectPartitionIds is null and empty for hive and other tables
```
 // see OptExternalPartitionPruner#computePartitionInfo:
 // it's not the same meaning when selectPartitionIds is null and empty for hive and other tables
```